api/src/main/java/jakarta/data/Pattern.java

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright (c) 2024 Contributors to the Eclipse Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package jakarta.data;
+
+import jakarta.data.metamodel.Restriction;
+
+
+/**
+ * Represents a pattern-based restriction for matching operations, encapsulating different
+ * options such as prefix, suffix, and substring matching. This implementation
+ * allows flexibility in creating pattern-based conditions directly as {@link Restriction} instances.
+ *
+ * <p>Example usage with metadata attributes:</p>
+ * <pre>
+ * Restriction<Book> prefixIgnoreCase = _Book.title.endsWith("Guide");
+ * </pre>
+ *
+ */
+public record Pattern(String value, boolean ignoreCase) {
+
+    /**
+     * Creates a pattern for a {@link Operator#LIKE LIKE} match on the specified value.
+     * This method sets the field to `null`, allowing it to be applied
+     * later to a specific attribute.
+     *
+     * @return a Pattern instance for a pattern match.
+     */
+    public static Pattern like(String pattern) {
+        return new Pattern(pattern, false);
+    }
+
+    /**
+     * Creates a pattern for a `LIKE` match where values start with the specified prefix.
+     * The `field` is set to `null` initially, allowing it to be assigned to an attribute later.
+     *
+     * @param value the prefix to match at the beginning of the field's value.
+     * @return a Pattern instance for a prefix match.
+     */
+    public static Pattern startsWith(String value) {
+        return new Pattern( value + "%", false);
```

Comment:
In my `Pattern` class, I `escape()` the `prefix`, because I figured it shouldn't be a pattern.

```java
    public static Pattern startsWith(String prefix) {
        return new Pattern(escape(prefix) + '%', true);
    }
```

Well, I mean, `startsWith()` is not a pattern in Java. I now notice that it *is* a pattern in the query by method name extension, so now I'm not quite sure what to do.

I suppose we should have `startsWIthLiteral()` / `containsLiteral()`, etc.

Reply:
I would leave StartsWith/EndsWith/Contains from Query by Method Name alone, not have them support Pattern, and not add additional methods for StartsWithLiteral..., taking only the minimal step of having Query by Method Name support Pattern for [Not]Like[IgnoreCase], which will sufficiently cover all of the operations.

We should consider if we want to change StartsWith/EndsWith/Contains to use literals rather than patterns in Restrictions and Parameter-based Automatic Methods, which are new in the current version.

Reply:
@gavinking sorry, I did not find your `Pattern` could you either give me or open a PR against this one?

Reply:
Never mind, I find it.

Reply:
I used your class and put the javadoc.